### ISSUE TITLE
When using CommonsChunkPlugin, the common bundle built contains hash list of all entry points

### ISSUE THREAD

**Issue by happypoulp** [webpack/webpack#660]
Hi,

First of all, thanks for the awesomeness that webpack is. I've been using it for few weeks only but I found that it provides all I expect from a bundler (and that I did not found anywhere else).

The current issue is the only thing that is currently preventing me from using it on production.

Here is the context:

I intend to use webpack for a big existing website to replace the current old-school bundle solution. This website is not a webapp and it is updated several times a day.

Our current bundle solution is built on a configuration file listing all bundles and the files they should embed. This allow us to have full control over what goes in a bundle and to maintain a common bundle for all pages. Then, each page is loaded with the common bundle and a page-specific bundle.

I tried to achieve the same with webpack and I used this configuration file:

```javascript
var webpack = require('webpack'),
    path = require('path');

var configuration =
    {
        entry:
        {
            page1: './js/page1.js',
            page2: './js/page2.js',
            page3: './js/page3.js',
        },
        output:
        {
            path: 'build',
            publicPath: 'build/',
            filename: '[name].[chunkhash].js',
            chunkFilename: '[id].[chunkhash].js'
        },
        plugins: [
            new webpack.NewWatchingPlugin(),
            new webpack.optimize.CommonsChunkPlugin('common.[chunkhash].js'),
            function()
            {
                this.plugin("done", function(stats)
                {
                    require("fs").writeFileSync(
                        path.join(__dirname, 'webpack_files_map.conf.json'),
                        JSON.stringify(stats.toJson().assetsByChunkName)
                    );
                });
            }
        ],
        resolve:
        {
            root: path.join(__dirname, 'js')
        }
    };

module.exports = configuration;
```

This configuration file produce correctly one bundle per page and a common bundle.

The behavior that is bothering me is that the common bundle embed the hash for all entry points:

```javascript
/******/ 	__webpack_require__.e = function requireEnsure(chunkId, callback) {
...
/******/ 			script.src = __webpack_require__.p + "" + chunkId + "." + {"0":"ec91e42ab36483e18cd9","1":"73705fff499a2fecdeeb","2":"5d017c0e038c150db29b"}[chunkId] + ".js";
```

This line of the common bundle make the common bundle implicitly dependent of ALL others bundles. So a modification of a file in page1.js that is not present in the common bundle will surely invalidate the common bundle because of the hash of page1.js that is written in the common bundle.

This may seems like a detail but the true purpose of a common bundle is to have a bundle that is long-cached and that do not change too often. But making this common bundle dependent of all others bundle kind of ruin this long-term strategy.

What I'd like is a way to specify this hashes externally (through a global var exposed in window) so that the common bundle is not dependent of others bundle addition / modification / deletion.

The code above could then become:

```javascript
/******/ 	__webpack_require__.e = function requireEnsure(chunkId, callback) {
...
/******/ 			script.src = __webpack_require__.p + "" + chunkId + "." + WEBPACK_BUNDLE_HASHES + [chunkId] + ".js";
```

The WEBPACK_BUNDLE_HASHES content could be extracted from the JSON stats file that is produced at the end of the compilation phase (I am not sure if this is already possible).

Do you think that this makes sense and that it could be done?

Thanks.

**Comment by sokra** [comment]
Makes sense. This is how soundcloud does it. 

You could do it this way:

``` js
        plugins: [
            new webpack.NewWatchingPlugin(),
            new webpack.optimize.CommonsChunkPlugin('common.[chunkhash].js'),
            new webpack.optimize.CommonsChunkPlugin('inline.js'),
            function()
```

and inline the content of `inline.js` into the HTML page as `<script>` tag.

---

For long time caching with `[chunkhash]` you should also use the `records` option, to make sure files don't change too often...

**Comment by sokra** [comment]
cc @diurnalist

**Comment by diurnalist** [comment]
@happypoulp yes, I had this same concern, and created the (perhaps badly-named) [chunk-manifest](https://github.com/diurnalist/chunk-manifest-webpack-plugin) plugin as a result. I should note that I apparently didn't write tests for that plugin yet - boo on me for that. But, maybe it will be helpful for you/others.

It will output all the chunk hashes to a JSON file that you can consume and use however you want. You just need to set it to some global variable (configurable) so webpack can read it at runtime.

@sokra your idea to use two CommonsChunkPlugin instances is interesting though. I'll have to see what that would look like, not sure if it's worth the extra bytes on the initial request or not. And, +1 to using the records option to keep module ids in sync across builds.

**Comment by happypoulp** [comment]
Thanks for the quick reply!

@sokra I tried your solution and even if it does extract hashes as desired, it came with the cost of a "big" inline.js because of the inclusion in it of the webpackBootstrap code. I'd like to be able to inline only hashes of bundles.

@diurnalist your solution is a perfect match for what I am looking for. The json produced is small and easily embeddable and the common plugin is stripped of other bundles hashes.

I also added the recordsPath options following your advices and I see a more consistent bundles update that I had before.

I'll go with @diurnalist solution and close the issue, thanks a lot!
